src/SonarQube.Client/Api/SonarQubeService.cs

```diff
@@ -0,0 +1,252 @@
+        {
+            httpClient = new HttpClient(messageHandler)
+            {
+                BaseAddress = connection.ServerUri,
+                DefaultRequestHeaders =
+                {
+                    Authorization = AuthenticationHeaderFactory.Create(
+                        connection.UserName, connection.Password, connection.Authentication),
+                },
+            };
+
+            httpClient.DefaultRequestHeaders.Add("User-Agent", userAgent);
+
+            IsConnected = true;
+
+            var versionResponse = await InvokeRequestAsync<IGetVersionRequest, string>(token);
+            sonarQubeVersion = Version.Parse(versionResponse);
+
+            var credentialResponse = await InvokeRequestAsync<IValidateCredentialsRequest, bool>(token);
+            if (!credentialResponse)
+            {
+                IsConnected = false;
+                throw new InvalidOperationException("Invalid credentials");
+            }
+        }
+
+        public void Disconnect()
+        {
+            IsConnected = false;
+            httpClient.Dispose();
+        }
+
+        private void EnsureIsConnected()
+        {
+            if (!IsConnected)
+            {
+                throw new InvalidOperationException("This operation expects the service to be connected.");
+            }
+        }
+
+        public async Task<IList<SonarQubeOrganization>> GetAllOrganizationsAsync(CancellationToken token) =>
+            await InvokeRequestAsync<IGetOrganizationsRequest, SonarQubeOrganization[]>(
+                request =>
+                {
+                    request.OnlyUserOrganizations = true;
+                },
+                token);
+
+        public async Task<IList<SonarQubePlugin>> GetAllPluginsAsync(CancellationToken token) =>
+            await InvokeRequestAsync<IGetPluginsRequest, SonarQubePlugin[]>(token);
+
+        public async Task<IList<SonarQubeProject>> GetAllProjectsAsync(string organizationKey, CancellationToken token) =>
+            await InvokeRequestAsync<IGetProjectsRequest, SonarQubeProject[]>(
+                request =>
+                {
+                    request.OrganizationKey = organizationKey;
+                },
+                token);
+
+        public async Task<IList<SonarQubeProperty>> GetAllPropertiesAsync(CancellationToken token) =>
+            await InvokeRequestAsync<IGetPropertiesRequest, SonarQubeProperty[]>(token);
+
+        public Uri GetProjectDashboardUrl(string projectKey)
+        {
+            EnsureIsConnected();
+
+            const string ProjectDashboardRelativeUrl = "dashboard/index/{0}";
+
+            return new Uri(httpClient.BaseAddress, string.Format(ProjectDashboardRelativeUrl, projectKey));
+        }
+
+        public async Task<SonarQubeQualityProfile> GetQualityProfileAsync(string projectKey, string organizationKey, SonarQubeLanguage language, CancellationToken token)
+        {
```

Comment:
Is it ok for any of the parameters to be null/empty?

Reply:
It is reasonable to at least log when there is organization key while SQ does not support organizations, because this would mean that we have inconsistency in our data... This is applicable for most of the API calls and I would prefer to implement this as a separate PR/ticket.

Reply:
#663